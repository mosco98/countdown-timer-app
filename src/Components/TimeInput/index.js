import React from 'react'

const TimeInput = ({
  timePaused,
  timeStarted,
  startTimeHandler,
  timeResetHandler,
  timeUp,
  startNewHandler,
  minuteInputHandler,
  minuteInput
}) => {
  return (
    <div className="d-flex align-items-center justify-content-center time-input-ctn ml-auto mr-auto w-50">
      {timeUp ? (
        <button
          className="btn btn-sm btn-success"
          onClick={startNewHandler}
          title="Start new countdown"
          data-testid="start-new-btn">
          Start new countdown
        </button>
      ) : (
        <React.Fragment>
          {timeStarted ? null : (
            <>
              <span className="text-black-50 font-weight-bold">Countdown:</span>
              <input
                className="form-control time-input ml-4 mr-4"
                type="number"
                placeholder="(Min)"
                name="minuteInput"
                onChange={minuteInputHandler}
                min="0"
                data-testid="minute-input"
              />
            </>
          )}
          {timePaused && (
            <button
              className="btn btn-sm btn-danger reset-btn"
              onClick={timeResetHandler}
              title="Reset timer"
              data-testid="reset-btn">
              Reset
            </button>
          )}
          {!timeStarted && (
            <button
              className={minuteInput ? 'btn btn-sm btn-success start-btn' : 'btn btn-sm disabled start-btn'}
              onClick={minuteInput ? startTimeHandler : null}
              title={minuteInput && 'Start timer'}
              data-testid="start-btn">
              Start
            </button>
          )}
        </React.Fragment>
      )}
    </div>
  )
}

export default TimeInput
