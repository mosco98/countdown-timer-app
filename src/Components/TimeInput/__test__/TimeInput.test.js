import { fireEvent, render } from '@testing-library/react'
import React from 'react'
import renderer from 'react-test-renderer'

import TimeInput from '..'

test('Initial Rendering', () => {
  render(<TimeInput />)
})

describe('Input value', () => {
  it('updates on change', () => {
    const { queryByTestId } = render(<TimeInput />)
    const timeInput = queryByTestId('minute-input')
    fireEvent.change(timeInput, { target: { value: '0' } })
    expect(timeInput.value).toBe('0')
  })
})

describe('Buttons', () => {
  it('start button renders correctly', () => {
    const props = {
      timeStarted: false,
      timePaused: false,
      timeUp: false
    }
    const { timeStarted, timePaused, timeUp } = props
    const component = renderer.create(<TimeInput timePaused={timePaused} timeStarted={timeStarted} timeUp={timeUp} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('reset button renders correctly', () => {
    const props = {
      timeStarted: true,
      timePaused: true,
      timeUp: false
    }
    const { timeStarted, timePaused, timeUp } = props
    const component = renderer.create(<TimeInput timePaused={timePaused} timeStarted={timeStarted} timeUp={timeUp} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('start-new-counter button renders correctly', () => {
    const props = {
      timeStarted: false,
      timePaused: false,
      timeUp: true
    }
    const { timeStarted, timePaused, timeUp } = props
    const component = renderer.create(<TimeInput timePaused={timePaused} timeStarted={timeStarted} timeUp={timeUp} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('Buttons functionality', () => {
  it('start button does not triggers startTimeHandler function when clicked', () => {
    const startTimeHandler = jest.fn()
    const { queryByTestId } = render(<TimeInput startTimeHandler={startTimeHandler} minuteInput={0} />)
    fireEvent.click(queryByTestId('start-btn'))
    expect(startTimeHandler).not.toHaveBeenCalled()
  })
  it('start button triggers startTimeHandler function when clicked', () => {
    const startTimeHandler = jest.fn()
    const num = (value) => {
      if (value !== '' || value > 0 || value % 1 !== 0) {
        return value
      }
    }
    const { queryByTestId } = render(<TimeInput startTimeHandler={startTimeHandler} minuteInput={() => num()} />)
    fireEvent.click(queryByTestId('start-btn'))
    expect(startTimeHandler).toHaveBeenCalled()
  })
  it('reset button triggers timeResetHandler function when clicked', () => {
    const timeResetHandler = jest.fn()
    const { queryByTestId } = render(<TimeInput timeResetHandler={timeResetHandler} timePaused={true} />)
    fireEvent.click(queryByTestId('reset-btn'))
    expect(timeResetHandler).toHaveBeenCalled()
  })
  it('start-new-timer button triggers startNewHandler function when clicked', () => {
    const startNewHandler = jest.fn()
    const { queryByTestId } = render(<TimeInput startNewHandler={startNewHandler} timeUp={true} />)
    fireEvent.click(queryByTestId('start-new-btn'))
    expect(startNewHandler).toHaveBeenCalled()
  })
})
