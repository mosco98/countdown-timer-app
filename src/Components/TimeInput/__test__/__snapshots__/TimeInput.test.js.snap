// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons reset button renders correctly 1`] = `
<div
  className="d-flex align-items-center justify-content-center time-input-ctn ml-auto mr-auto w-50"
>
  <button
    className="btn btn-danger reset-btn"
    data-testid="reset-btn"
    title="Reset timer"
  >
    Reset
  </button>
</div>
`;

exports[`Buttons start button renders correctly 1`] = `
<div
  className="d-flex align-items-center justify-content-center time-input-ctn ml-auto mr-auto w-50"
>
  <span
    className="text-black-50 font-weight-bold"
  >
    Countdown:
  </span>
  <input
    className="form-control time-input ml-4 mr-4"
    data-testid="minute-input"
    min="0"
    name="minuteInput"
    placeholder="(Min)"
    type="number"
  />
  <button
    className="btn btn-sm disabled start-btn"
    data-testid="start-btn"
    onClick={null}
  >
    Start
  </button>
</div>
`;

exports[`Buttons start-new-counter button renders correctly 1`] = `
<div
  className="d-flex align-items-center justify-content-center time-input-ctn ml-auto mr-auto w-50"
>
  <button
    className="btn btn-block btn-success"
    data-testid="start-new-btn"
    title="Start new countdown"
  >
    Start new countdown
  </button>
</div>
`;
